### ISSUE TITLE
[BUG] Issue of the class ForecastingHorizon and related functions causing downstream issues with Prophet Ensemble

### ISSUE THREAD

**Issue by ngupta23** [alan-turing-institute/sktime#561]
**Describe the bug**
Since Prophet needs the index to be in `pd.DatetimeIndex` format, it does not work in ensemble mode with other models.

**To Reproduce**

```python
y = load_airline()
y_train, y_test = temporal_train_test_split(y, test_size=36)
fh = ForecastingHorizon(y_test.index, is_relative=False)

forecaster = EnsembleForecaster([
    ("autoARIMA", AutoARIMA(sp=12, suppress_warnings=True)),
    ("autoETS", AutoETS(auto=True, sp=12, n_jobs=-1)),
    ("fbprophet", Prophet(seasonality_mode="multiplicative", add_country_holidays={"country_name": "US"}))
])
%time forecaster.fit(y_train)
y_pred = forecaster.predict(fh)
```

Error message is as follows
```terminal
NotImplementedError: <class 'pandas.core.indexes.period.PeriodIndex'> is not supported. Please use type: <class 'pandas.core.indexes.datetimes.DatetimeIndex'> instead.
```

**Expected behavior**
It would be ok to provide a working example of this use case. If it means I have to change data index to `pd.DatetimeIndex` for other models, I am ok doing that but I tried it and it gave an error as well

```python
# Convert index to pd.DatetimeIndex
z = y.copy()
z.index = y.index.to_timestamp()
z_train, z_test = temporal_train_test_split(z, test_size=36)

forecaster = EnsembleForecaster([
    ("autoARIMA", AutoARIMA(sp=12, suppress_warnings=True)),
    ("autoETS", AutoETS(auto=True, sp=12, n_jobs=-1)),
    ("fbprophet", Prophet( seasonality_mode="multiplicative", add_country_holidays={"country_name": "US"}))
])
%time forecaster.fit(z_train)
y_pred = forecaster.predict(fh)
```

Error Message is as follows

```terminal
sktime\forecasting\base\_fh.py in _check_cutoff(self, cutoff)
    388 
    389         if isinstance(index, pd.PeriodIndex):
--> 390             assert isinstance(cutoff, pd.Period)
    391         if isinstance(index, pd.DatetimeIndex):
    392             assert isinstance(cutoff, pd.Timestamp)
```

I tried another variant of the above (changes the arg to the pred function) but got a different error

```python
forecaster = EnsembleForecaster([
    ("autoARIMA", AutoARIMA(sp=12, suppress_warnings=True)),
    ("autoETS", AutoETS(auto=True, sp=12, n_jobs=-1)),
    ("fbprophet", Prophet( seasonality_mode="multiplicative", add_country_holidays={"country_name": "US"}))
])
%time forecaster.fit(z_train)
y_pred = forecaster.predict(fh.to_relative(cutoff=y_train.index[-1]))
```

```
sktime\forecasting\base\_fh.py in _check_values(values)
    108     # check values does not contain duplicates
    109     if len(values) != values.nunique():
--> 110         raise ValueError("`values` must not contain duplicates.")
    111 
    112     # return sorted values

ValueError: `values` must not contain duplicates.
```

**Versions**
System:
    python: 3.6.12 |Anaconda, Inc.| (default, Sep  9 2020, 00:29:25) [MSC v.1916 64 bit (AMD64)]
executable: C:\Users\Nikhil\.conda\envs\sktime_dev\python.exe
   machine: Windows-10-10.0.18362-SP0

Python dependencies:
          pip: 20.3.3
   setuptools: 51.0.0.post20201207
      sklearn: 0.24.0
       sktime: 0.5.0
  statsmodels: 0.12.1
        numpy: 1.19.4
        scipy: 1.5.4
       Cython: 0.29.17
       pandas: 1.1.5
   matplotlib: 3.3.3
       joblib: 1.0.0
        numba: 0.52.0
     pmdarima: 1.8.0
      tsfresh: 0.17.0

**Comment by aiwalter** [comment]
@ngupta23 is seems you are using a `fh` with `PeriodIndex`.
> ```python
>fh = ForecastingHorizon(y_test.index, is_relative=False)
>```
Pls try to convert it also to `DatetimeIndex` or use a relative `fh`. Does it work then?

**Comment by aiwalter** [comment]
@ngupta23 

> I tried another variant of the above (changes the arg to the pred function) but got a different error

This bug is already known, as descibed in #534

**Comment by ngupta23** [comment]
Yes, I tried this variant now and am getting a similar error (I am running on dev version from github, not PyPi version). I am still trying to get used to this concept of `ForecastingHorizon` class, its various options and when to use which one so thanks for your understanding as I work through this.

This is my new code variant snippets (missing parts are same as before). Can you provide me the right syntax that you recommend to use once all the underlying bugs are resolved.

### Version 1

* `fh` is `pd.DateTimeIndex`
* `train` used z_train (from `forecasting.ipynb` example notebook)
* `fit` uses `fh` converted to relative values (from `forecasting.ipynb` example notebook)

```python
fh = ForecastingHorizon(y_test.index.to_timestamp(), is_relative=False)

forecaster = EnsembleForecaster([
    ("autoARIMA", AutoARIMA(sp=12, suppress_warnings=True)),
    ("autoETS", AutoETS(auto=True, sp=12, n_jobs=-1)),
    ("fbprophet", Prophet(seasonality_mode="multiplicative", add_country_holidays={"country_name": "US"}))  # Does not work yet
])
%time forecaster.fit(z_train)
y_pred = forecaster.predict(fh.to_relative(cutoff=y_train.index[-1]))
```

```python
----> 8 y_pred = forecaster.predict(fh.to_relative(cutoff=y_train.index[-1]))
     
C:\Users\Nikhil\sktime_forked\sktime\forecasting\base\_fh.py in to_relative(self, cutoff)
    220 
    221         else:
--> 222             self._check_cutoff(cutoff)
    223             values = self.to_pandas() - cutoff
    224 

C:\Users\Nikhil\sktime_forked\sktime\forecasting\base\_fh.py in _check_cutoff(self, cutoff)
    390             assert isinstance(cutoff, pd.Period)
    391         if isinstance(index, pd.DatetimeIndex):
--> 392             assert isinstance(cutoff, pd.Timestamp)

AssertionError:
```

### Variant 2
* `fh` is `pd.DateTimeIndex`
* `train` used z_train (from `forecasting.ipynb` example notebook)
* `fit` uses `fh` directly

```python
fh = ForecastingHorizon(y_test.index.to_timestamp(), is_relative=False)

forecaster = EnsembleForecaster([
    ("autoARIMA", AutoARIMA(sp=12, suppress_warnings=True)),
    ("autoETS", AutoETS(auto=True, sp=12, n_jobs=-1)),
    ("fbprophet", Prophet(seasonality_mode="multiplicative", add_country_holidays={"country_name": "US"}))  # Does not work yet
])
%time forecaster.fit(z_train)
y_pred = forecaster.predict(fh)
```

```python
      6 ])
      7 get_ipython().magic('time forecaster.fit(z_train)')
----> 8 y_pred = forecaster.predict(fh)

C:\Users\Nikhil\sktime_forked\sktime\forecasting\base\_sktime.py in predict(self, fh, X, return_pred_int, alpha)
    225         self.check_is_fitted()
    226         self._set_fh(fh)
--> 227         return self._predict(self.fh, X, return_pred_int=return_pred_int, alpha=alpha)
    228 
    229     def compute_pred_int(self, y_pred, alpha=DEFAULT_ALPHA):

C:\Users\Nikhil\sktime_forked\sktime\forecasting\compose\_ensemble.py in _predict(self, fh, X, return_pred_int, alpha)
     76         if return_pred_int:
     77             raise NotImplementedError()
---> 78         return pd.concat(self._predict_forecasters(fh, X), axis=1).mean(axis=1)

C:\Users\Nikhil\sktime_forked\sktime\forecasting\base\_meta.py in _predict_forecasters(self, fh, X, return_pred_int, alpha)
     78         return [
     79             forecaster.predict(fh, X, return_pred_int=return_pred_int, alpha=alpha)
---> 80             for forecaster in self.forecasters_
     81         ]
     82 

C:\Users\Nikhil\sktime\forecasting\base\_meta.py in <listcomp>(.0)
     78         return [
     79             forecaster.predict(fh, X, return_pred_int=return_pred_int, alpha=alpha)
---> 80             for forecaster in self.forecasters_
     81         ]
     82 

C:\Users\Nikhil\sktime\forecasting\base\_sktime.py in predict(self, fh, X, return_pred_int, alpha)
    225         self.check_is_fitted()
    226         self._set_fh(fh)
--> 227         return self._predict(self.fh, X, return_pred_int=return_pred_int, alpha=alpha)
    228 
    229     def compute_pred_int(self, y_pred, alpha=DEFAULT_ALPHA):

C:\Users\Nikhil\sktime_forked\sktime\forecasting\arima.py in _predict(self, fh, X, return_pred_int, alpha)
    370     def _predict(self, fh, X=None, return_pred_int=False, alpha=DEFAULT_ALPHA):
    371         # distinguish between in-sample and out-of-sample prediction
--> 372         fh_oos = fh.to_out_of_sample(self.cutoff)
    373         fh_ins = fh.to_in_sample(self.cutoff)
    374 

C:\Users\Nikhil\sktime_forked\sktime\forecasting\base\_fh.py in to_out_of_sample(self, cutoff)
    303             Out-of-sample values of forecasting horizon
    304         """
--> 305         is_out_of_sample = self._is_out_of_sample(cutoff)
    306         out_of_sample = self.to_pandas()[is_out_of_sample]
    307         return self._new(out_of_sample)

C:\Users\Nikhil\sktime_forked\sktime\forecasting\base\_fh.py in _is_out_of_sample(self, cutoff)
    331         """Get index location of out-of-sample values"""
    332         # return ~self._in_sample_idx(cutoff)
--> 333         return self.to_relative(cutoff).to_pandas() > 0
    334 
    335     def is_all_out_of_sample(self, cutoff=None):

C:\Users\Nikhil\sktime_forked\sktime\forecasting\base\_fh.py in to_relative(self, cutoff)
    226                 values = _coerce_duration_to_int(values, unit=_get_unit(cutoff))
    227 
--> 228             return self._new(values, is_relative=True)
    229 
    230     @lru_cache(typed=True)

C:\Users\Nikhil\sktime_forked\sktime\forecasting\base\_fh.py in _new(self, values, is_relative)
    163         if is_relative is None:
    164             is_relative = self.is_relative
--> 165         return type(self)(values, is_relative)
    166 
    167     @property

C:\Users\Nikhil\sktime\forecasting\base\_fh.py in __init__(self, values, is_relative)
    140         if not isinstance(is_relative, bool):
    141             raise TypeError("`is_relative` must be a boolean")
--> 142         values = _check_values(values)
    143 
    144         # check types, note that isinstance() does not work here because index

C:\Users\Nikhil\sktime_forked\sktime\forecasting\base\_fh.py in _check_values(values)
    108     # check values does not contain duplicates
    109     if len(values) != values.nunique():
--> 110         raise ValueError("`values` must not contain duplicates.")
    111 
    112     # return sorted values

ValueError: `values` must not contain duplicates.
```

**Comment by aiwalter** [comment]
@ngupta23 
I was playing a bit around and was able to recreate this. This issues do not come from `from sktime.forecasting.fbprophet import Prophet`. Could you pls rename the this issue? It seems to be solely an issue of the class `ForecastingHorizon` and related functions.

**Comment by mloning** [comment]
Yes same here. It's not even the pd.DatetimeIndex by itself but the combination with monthly ("MS") a few other frequencies that's causing the problem. 

@ngupta23 this for example works:

```python
from sktime.forecasting.all import *
from sktime.forecasting.ets import AutoETS
from sktime.forecasting.fbprophet import Prophet
from sktime.utils._testing.forecasting import make_forecasting_problem

y = make_forecasting_problem(n_timepoints=100)  # toy data with daily frequency
y_train, y_test = temporal_train_test_split(y, test_size=36)

fh = ForecastingHorizon(y_test.index, is_relative=False)
forecaster = EnsembleForecaster([
    ("autoARIMA", AutoARIMA(sp=12, suppress_warnings=True)),
    ("autoETS", AutoETS(auto=True, sp=12, n_jobs=-1)),
    ("fbprophet", Prophet(seasonality_mode="multiplicative", add_country_holidays={"country_name": "US"}))
])
forecaster.fit(y_train)
y_pred = forecaster.predict(fh)
```

**Comment by mloning** [comment]
Closed in favour or #534
